api/search_test.go

```diff
@@ -6,250 +6,241 @@ import (
+		req := httptest.NewRequest("GET", "http://localhost:8080/search?limit=a", nil)
 		resp := httptest.NewRecorder()
 
 		searchHandler.ServeHTTP(resp, req)
 
 		So(resp.Code, ShouldEqual, http.StatusBadRequest)
-		So(resp.Body.String(), ShouldContainSubstring, "Invalid size parameter")
+		So(resp.Body.String(), ShouldContainSubstring, "Invalid limit parameter")
+		So(qbMock.BuildSearchQueryCalls(), ShouldHaveLength, 0)
 		So(esMock.MultiSearchCalls(), ShouldHaveLength, 0)
 	})
 
-	Convey("Should return BadRequest for invalid from parameter", t, func() {
-		setupSearchTestTemplates("dummy")
-		esMock := &ElasticSearcherMock{
-			MultiSearchFunc: func(ctx context.Context, index string, docType string, request []byte) ([]byte, error) {
-				return nil, nil
-			},
-		}
+	Convey("Should return BadRequest for invalid offset parameter", t, func() {
+		qbMock := newQueryBuilderMock(nil, nil)
+		esMock := newElasticSearcherMock(nil, nil)
+		trMock := newResponseTransformerMock(nil, nil)
 
-		searchHandler := SearchHandlerFunc(esMock)
+		searchHandler := SearchHandlerFunc(qbMock, esMock, trMock)
 
-		req := httptest.NewRequest("GET", "http://localhost:8080/search?from=b", nil)
+		req := httptest.NewRequest("GET", "http://localhost:8080/search?offset=b", nil)
 		resp := httptest.NewRecorder()
 
 		searchHandler.ServeHTTP(resp, req)
 
 		So(resp.Code, ShouldEqual, http.StatusBadRequest)
-		So(resp.Body.String(), ShouldContainSubstring, "Invalid from parameter")
+		So(resp.Body.String(), ShouldContainSubstring, "Invalid offset parameter")
+		So(qbMock.BuildSearchQueryCalls(), ShouldHaveLength, 0)
 		So(esMock.MultiSearchCalls(), ShouldHaveLength, 0)
 	})
 
-	Convey("Should return InternalError for invalid template", t, func() {
-		setupSearchTestTemplates("dummy{{.Moo}}")
-		esMock := &ElasticSearcherMock{
-			MultiSearchFunc: func(ctx context.Context, index string, docType string, request []byte) ([]byte, error) {
-				return nil, nil
-			},
-		}
+	// Should return Internal Error for error from query builder
+
+	Convey("Should return InternalError for errors returned from query builder", t, func() {
+		qbMock := newQueryBuilderMock(nil, errors.New("Something"))
+		esMock := newElasticSearcherMock(nil, nil)
+		trMock := newResponseTransformerMock(nil, nil)
 
-		searchHandler := SearchHandlerFunc(esMock)
+		searchHandler := SearchHandlerFunc(qbMock, esMock, trMock)
 
-		req := httptest.NewRequest("GET", "http://localhost:8080/search", nil)
+		req := httptest.NewRequest("GET", "http://localhost:8080/search?q=a", nil)
 		resp := httptest.NewRecorder()
 
 		searchHandler.ServeHTTP(resp, req)
 
 		So(resp.Code, ShouldEqual, http.StatusInternalServerError)
-		So(resp.Body.String(), ShouldContainSubstring, "Failed to create query")
+		So(resp.Body.String(), ShouldContainSubstring, "Failed to create search query")
+		So(qbMock.BuildSearchQueryCalls(), ShouldHaveLength, 1)
+		So(qbMock.BuildSearchQueryCalls()[0].Q, ShouldResemble, "a")
```

Comment:
The valid query is a valid generated elastic query which has been built (mocked) by the queryBuilder. This is different from the query parameter `q`. It may be worth a different constant to replace the literal 'a' in order to clarify this.